packages/electron/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,441 @@
+        // Set the electron version for both the dev and the production mode. (https://github.com/eclipse-theia/theia/issues/3254)
+        // Otherwise, the forked backend processes will not know that they're serving the electron frontend.
+        process.env.THEIA_ELECTRON_VERSION = process.versions.electron;
+        if (noBackendFork) {
+            process.env[ElectronSecurityToken] = JSON.stringify(this.electronSecurityToken);
+            // The backend server main file is supposed to export a promise resolving with the port used by the http(s) server.
+            const address: AddressInfo = await require(this.globals.THEIA_BACKEND_MAIN_PATH);
+            return address.port;
+        } else {
+            const backendProcess = fork(this.globals.THEIA_BACKEND_MAIN_PATH, backendArgv, await this.getForkOptions());
+            return new Promise((resolve, reject) => {
+                // The backend server main file is also supposed to send the resolved http(s) server port via IPC.
+                backendProcess.on('message', (address: AddressInfo) => {
+                    resolve(address.port);
+                });
+                backendProcess.on('error', error => {
+                    reject(error);
+                });
+                app.on('quit', () => {
+                    // If we forked the process for the clusters, we need to manually terminate it.
+                    // See: https://github.com/eclipse-theia/theia/issues/835
+                    process.kill(backendProcess.pid);
+                });
+            });
+        }
+    }
+
+    protected async getForkOptions(): Promise<ForkOptions> {
+        return {
+            env: {
+                ...process.env,
+                [ElectronSecurityToken]: JSON.stringify(this.electronSecurityToken),
+            },
+        };
+    }
+
+    protected async attachElectronSecurityToken(port: number): Promise<void> {
+        await new Promise((resolve, reject) => {
+            electron.session.defaultSession!.cookies.set({
+                url: `http://localhost:${port}`,
+                name: ElectronSecurityToken,
+                value: JSON.stringify(this.electronSecurityToken),
+                httpOnly: true,
+            }, error => error ? reject(error) : resolve());
+        });
+    }
+
+    protected hookApplicationEvents(): void {
+        app.on('will-quit', this.onWillQuit.bind(this));
+        app.on('second-instance', this.onSecondInstance.bind(this));
+        app.on('window-all-closed', this.onWindowAllClosed.bind(this));
+    }
+
+    protected onWillQuit(event: ElectronEvent): void {
+        this.stopContributions();
+    }
+
+    protected async onSecondInstance(event: ElectronEvent, argv: string[], cwd: string): Promise<void> {
+        await this.launch({ argv, cwd, secondInstance: true });
+    }
+
+    protected onWindowAllClosed(event: ElectronEvent): void {
+        this.requestStop();
```

Comment:
@kittaakos I aligned the handling of the event with the others handlers so that it's easier to extend.